### PR TITLE
simplify `Vec` and `Point` constructors

### PR DESCRIPTION
**Pull request by ehgus**
follow-up of #601 

Type promotion of `Point` is rely on `Vec`, so I remove the floating-point promotion at a `Point` constructor. `Vec` promotes type `T` into `float(T)`. I listed possible scenarios how the promotion will work:

|T| float(T) |
|---|---|
|Integer|Float64|
|Float64|Float64|
|Float32|Float32|
|Quantity{Int64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|Quantity{Float64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|
|Quantity{Float32, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|Quantity{Float32, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|
|Quantity{Float64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|Quantity{Float64, 𝐋, Unitful.FreeUnits{(m,), 𝐋, nothing}}|

The table proves that the `Unitful` quantities will be promoted as expected. I have checked with my test code:
```Julia repl
julia> using Meshes

julia> x= 1u"m"
1 m

julia> Point(x,x,x)
Point(1.0 m, 1.0 m, 1.0 m)

julia> y=1f0u"m"
1.0f0 m

julia> Point(x,y,x)
Point(1.0f0 m, 1.0f0 m, 1.0f0 m)
```

### CODE REVIEW

File: src/primitives/point.jl
Comment (juliohm) [inline src/primitives/point.jl:43]: Do we need this inner constructor or is it automatically generated by the Julia compiler? Can we try to erase it completely here?

File: src/vectors.jl
Comment (juliohm) [inline src/vectors.jl]: Can we add type parameters for the inner constructor? I think it is the default, we never used inner constructors in Julia without these parameters.

**Review by juliohm — CHANGES_REQUESTED** [review]
This looks great @ehgus , I attached a few comments/suggestions based off a quick chat with @eliascarv. He will also try to review by the end of the day.

File: src/primitives/point.jl
Comment (ehgus) [reply src/primitives/point.jl:43]: The inner constructor is necessary. The constructor is the inference rule for interpreting the implicit `Point` type to the explicit `Point` type. The constructors below cannot replace this.

Comment (ehgus) [reply src/primitives/point.jl:43]: I deleted the line and tested the code, which failed some tests.

File: src/vectors.jl
Comment (ehgus) [reply src/vectors.jl]: The previous inner constructors presume that the type parameter of `Vec` is the same as that of `coords`.
Instead, my code infers the type parameter of `Vec`  from `coords`.
The proposed code will work like the previous one with more relaxed conditions.

Comment (eliascarv) [inline src/vectors.jl]: I think `NTuple{Dim,float(T)}` is not needed:
```suggestion
  Vec(coords::NTuple{Dim,T}) where {Dim,T} = new{Dim,float(T)}(coords)
```

File: src/vectors.jl
Comment (eliascarv) [inline src/vectors.jl]: With this, we can remove the `checkdim` function:
```suggestion
function Vec{Dim,T}(coords::Union{Tuple,AbstractVector}) where {Dim,T}
  if Dim ≠ length(coords)
    throw(DimensionMismatch("invalid dimension"))
  end
  Vec(NTuple{Dim,float(T)}(coords))
end
```
Please, along with this change remove the `checkdim` function.

**Review by eliascarv — CHANGES_REQUESTED** [review]

File: src/vectors.jl
Comment (eliascarv) [inline src/vectors.jl]: The `Vec` constructor must only support tuples.
```suggestion
Vec(coords::Tuple) = Vec(promote(coords...))
```

**Review by eliascarv — CHANGES_REQUESTED** [review]

**Comment by juliohm** [comment]
@ehgus can you please take a look at the updated suggestions?

File: src/primitives/point.jl
Comment (eliascarv) [inline src/primitives/point.jl]: I think we can simplify these constructors:
```julia
# convenience constructors
Point{Dim,T}(coords...) where {Dim,T} = Point(Vec{Dim,T}(coords...))
Point(coords...) = Point(Vec(coords...))
```

Comment (ehgus) [reply src/primitives/point.jl]: With the change,  `Point` constructors can no longer take `Tuple` as an argument.
I agree that construction like `Point((1,2,3))` may lead to unnecessary complexity in the code, and users can easily handle the change. I want to make sure that other maintainers also agree with that.

File: src/tolerances.jl
Comment (juliohm) [inline src/tolerances.jl]: If this change is unrelated to the Point constructor, we should probably add it in a separate PR.

File: src/primitives/point.jl
Comment (juliohm) [reply src/primitives/point.jl]: Which change @ehgus? The one proposed by @eliascarv above?

**Review by juliohm — CHANGES_REQUESTED** [review]
We need additional tests to make sure that the constructors work with units.

File: src/primitives/point.jl
Comment (eliascarv) [reply src/primitives/point.jl]: The constructor I proposed just forwards the arguments from the `Point(args...)` constructor to the `Vec(args...)` constructor. Therefore `Point((1, 2, 3))` would be equivalent to `Point(Vec((1, 2, 3)))`.

**Comment by eliascarv** [comment]
@ehgus, please remove the `checkdim` function:

image

This function is no longer needed in the PR code.

File: src/tolerances.jl
Comment (ehgus) [reply src/tolerances.jl]: Yes, this change should be in another PR. Thank you!

**Comment by juliohm** [comment]
Thank you @ehgus for improving the PR further. As soon as you add tests I can review and approve. Then, we can wait for @eliascarv 's review and merge it.

File: src/vectors.jl
Comment (ehgus) [inline src/vectors.jl:86]: > @ehgus, please remove the `checkdim` function:
> 
> image
> 
> This function is no longer needed in the PR code.

I removed the unused function

File: src/primitives/point.jl
Comment (ehgus) [reply src/primitives/point.jl]: I should have tested your code before commenting. Thank you @eliascarv. Now I understand your suggestion and have updated the code.

**Review by eliascarv — APPROVED** [review]

File: test/primitives.jl
Comment (juliohm) [inline test/primitives.jl]: Instead of checking the coordtype, what about just checking the `unit(coordinates(point))` and comparing with the expected value? The Unitful types are quite verbose and likely include implementation details that may change with time.

Comment (eliascarv) [reply test/primitives.jl]: Suggestion:
```julia
@test unit(coordtype(point)) == u"m"
@test Unitful.numtype(coordtype(point)) === Float64
```

Comment (ehgus) [reply test/primitives.jl]: Thank you for the suggestions, @eliascarv and @juliohm. I have been looking for a function like `Unitful.numtype`. I will soon update the test code.

**Comment by ehgus** [comment]
Tests for `Vec` are also added

**Review by juliohm — APPROVED** [review]

**Comment by juliohm** [comment]
Thank you @ehgus for this very important contribution. Feel free to submit another PR with the `atol` improvements we discussed.

**Pull request merged**
